### PR TITLE
[Merged by Bors] - Update `wgpu` to 0.14.0, `naga` to `0.10.0`, `winit` to 0.27.4, `raw-window-handle` to 0.5.0, `ndk` to 0.7

### PR DESCRIPTION
**Pull request by VitalyAnkh**
# Objective

- Update `wgpu` to 0.14.0, `naga` to `0.10.0`, `winit` to 0.27.4, `raw-window-handle` to 0.5.0, `ndk` to 0.7.

## Solution

---

### Changed

- Changed `RawWindowHandleWrapper` to `RawHandleWrapper` which wraps both `RawWindowHandle` and `RawDisplayHandle`, which satisfies the `impl HasRawWindowHandle and HasRawDisplayHandle` that `wgpu` 0.14.0 requires.

- Changed `bevy_window::WindowDescriptor`'s `cursor_locked` to `cursor_grab_mode`, change its type from `bool` to `bevy_window::CursorGrabMode`.

## Migration Guide

- Adjust usage of `bevy_window::WindowDescriptor`'s `cursor_locked` to `cursor_grab_mode`, and adjust its type from `bool` to `bevy_window::CursorGrabMode`.

### CODE REVIEW

**Comment by mockersf** [comment]
related to #5347 and #6086

File: crates/bevy_window/src/window.rs [hunk 1]
```diff
@@ -290,6 +289,18 @@ pub enum WindowCommand {
     Close,
 }
 
+/// Defines if and how the cursor is grabbed.
```
Comment (alice-i-cecile) [inline crates/bevy_window/src/window.rs:292]: This should briefly describe how this struct might be used.

This is easy to encounter on docs.rs, but doesn't hint at usage.

File: crates/bevy_winit/src/lib.rs [hunk 2]
```diff
@@ -727,3 +727,12 @@ fn handle_create_window_events(
         }
     }
 }
+
+/// Map `bevy_window::CursorGrabMode` to `winit::window::CursorGrabMode`.
+fn winit_grab_mode(mode: bevy_window::CursorGrabMode) -> winit::window::CursorGrabMode {
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_winit/src/lib.rs:732]: Or I guess an `Into` impl if orphan rules get in the way.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
A couple tiny issues, but I'd like to get moving on this one. Reviewers: please test this one out; there's lots of potential for new serious hardware / OS / driver issues.

File: deny.toml [hunk 3]
```diff
@@ -35,18 +35,10 @@ wildcards = "deny"
...
-    { name = "ndk", version = "0.5" },                  # from winit v0.26.1
-    { name = "ndk", version = "0.6" },                  # from rodio v0.16.0
-    { name = "ndk-glue", version = "0.5" },             # from winit v0.26.1
-    { name = "ndk-sys", version = "0.2" },              # from winit v0.26.1
     { name = "ndk-sys", version = "0.3" },              # from rodio v0.16.0
-    { name = "parking_lot", version = "0.11" },         # from winit v0.26.1
-    { name = "parking_lot_core", version = "0.8" },     # from winit v0.26.1
-    { name = "raw-window-handle", version = "0.4" },    # from wgpu v0.13.0
+    { name = "ndk", version = "0.7" },                  # from winit v0.27.4
+    { name = "ndk-glue", version = "0.7" },             # from winit v0.27.4
```
Comment (mockersf) [inline deny.toml:40]: This should skip the oldest version of those rather than the newest

File: crates/bevy_window/src/window.rs [hunk 4]
```diff
@@ -592,34 +603,34 @@ impl Window {
...
+    /// - **`macOS`** doesn't support cursor grab, but most windowing plugins can emulate it. See [issue #4875](https://github.com/bevyengine/bevy/issues/4875#issuecomment-1153977546) for more information.
     /// - **`iOS/Android`** don't have cursors.
     #[inline]
-    pub fn cursor_locked(&self) -> bool {
-        self.cursor_locked
+    pub fn cursor_grab_mode(&self) -> CursorGrabMode {
+        self.cursor_grab_mode
     }
-    /// Set whether or not the cursor is locked.
+    /// Set whether and how the cursor is grabed.
```
Comment (torsteingrindvik) [inline crates/bevy_window/src/window.rs:616]:
```suggestion
    /// Set whether and how the cursor is grabbed.
```

File: crates/bevy_window/src/window.rs [hunk 5]
```diff
@@ -592,34 +603,34 @@ impl Window {
         self.command_queue
             .push(WindowCommand::SetDecorations { decorations });
     }
-    /// Get whether or not the cursor is locked.
+    /// Get whether or how the cursor is grabed.
```
Comment (torsteingrindvik) [inline crates/bevy_window/src/window.rs:606]:
```suggestion
    /// Get whether or how the cursor is grabbed.
```

File: crates/bevy_window/src/window.rs (hunk 4 above)
Comment (VitalyAnkh) [reply crates/bevy_window/src/window.rs]: Thanks!

File: crates/bevy_winit/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (VitalyAnkh) [reply crates/bevy_winit/src/lib.rs:724]: I found it's not comfortable to do this for `bevy_window::CursorGrabMode` and `winit::window:CursorGrabMode` are not defined in the `bevy_winit` crate so orphan rules deny any `From` and `Into` impl. How about just using a `match` expression, removing this function?

[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_winit/src/lib.rs:724]: Sure I'm fine with that.

File: deny.toml (hunk 3 above)
Comment (VitalyAnkh) [reply deny.toml]: Done, thanks

File: crates/bevy_winit/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (VitalyAnkh) [reply crates/bevy_winit/src/lib.rs:724]: Using a simple match causes code duplication so I suggest leaving code as it is :smile:

File: crates/bevy_window/src/window.rs (hunk 1 above)
Comment (VitalyAnkh) [reply crates/bevy_window/src/window.rs:345]: Great. Added.

**Comment by VitalyAnkh** [comment]
I feel this PR is in good shape now. Ping @alice-i-cecile @mockersf.

**Comment by rparrett** [comment]
Tested all examples in the `window` folder.

**Review by alice-i-cecile — APPROVED** [review]
Code looks good, examples are running locally for me on Windows 10 laptop. This LGTM now; I'd like to merge this to unblock bug fixes and related features.

**Comment by devil-ira** [comment]
Tested all examples in the `window` folder and most of the ones in `shader` on Manjaro KDE.

File: crates/bevy_winit/src/lib.rs [hunk 6]
```diff
@@ -719,3 +719,12 @@ fn handle_create_window_events(
         }
     }
 }
+
+/// Map [`bevy_window::CursorGrabMode`] to [`winit::window::CursorGrabMode`].
+fn winit_grab_mode(mode: bevy_window::CursorGrabMode) -> winit::window::CursorGrabMode {
```
Comment (mockersf) [inline crates/bevy_winit/src/lib.rs:724]: could you put this with its friends in https://github.com/bevyengine/bevy/blob/main/crates/bevy_winit/src/converters.rs

File: deny.toml [hunk 7]
```diff
@@ -35,18 +35,10 @@ wildcards = "deny"
...
-    { name = "ndk", version = "0.5" },                  # from winit v0.26.1
-    { name = "ndk", version = "0.6" },                  # from rodio v0.16.0
-    { name = "ndk-glue", version = "0.5" },             # from winit v0.26.1
-    { name = "ndk-sys", version = "0.2" },              # from winit v0.26.1
     { name = "ndk-sys", version = "0.3" },              # from rodio v0.16.0
-    { name = "parking_lot", version = "0.11" },         # from winit v0.26.1
-    { name = "parking_lot_core", version = "0.8" },     # from winit v0.26.1
-    { name = "raw-window-handle", version = "0.4" },    # from wgpu v0.13.0
+    { name = "ndk", version = "0.6" },                  # from winit v0.26.0
+    { name = "ndk-glue", version = "0.6" },             # from winit v0.26.0
```
Comment (mockersf) [inline deny.toml:40]: do we still have a dependency on winit 0.26?

**Comment by mockersf** [comment]
bors try

File: examples/input/mouse_grab.rs [hunk 8]
```diff
@@ -1,6 +1,7 @@
 //! Demonstrates how to grab and hide the mouse cursor.
 
 use bevy::prelude::*;
+use bevy::window::CursorGrabMode;
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply examples/input/mouse_grab.rs:4]: IMO no; you only need to call this once.

**Comment by mockersf** [comment]
wasm screenshots in CI seem good 👍

File: crates/bevy_winit/src/lib.rs (hunk 6 above)
Comment (VitalyAnkh) [reply crates/bevy_winit/src/lib.rs:724]: Sure, didn't know this before :smile:

Comment (VitalyAnkh) [reply crates/bevy_winit/src/lib.rs]: Done.

**Comment by rparrett** [comment]
I don't know what the normal state of affairs is for iOS, but in the simulator I have working touch camera controls and sound, but the button does not work.

The console gets spammed with

> 2022-10-19T14:20:12.198663Z  WARN `bevy_render::camera::camera_driver_node`: Camera priority ambiguities detected for active cameras with the following priorities: {(0, Window(WindowId(00000000-0000-0000-0000-000000000000)))}. To fix this, ensure there is exactly one Camera entity spawned with a given priority for a given RenderTarget. Ambiguities should be resolved because either (1) multiple active cameras were spawned accidentally, which will result in rendering multiple instances of the scene or (2) for cases where multiple active cameras is intentional, ambiguities could result in unpredictable render results.

**Comment by rparrett** [comment]
The ios example is spawning 2 cameras for some reason.

**Comment by rparrett** [comment]
Looks like the button is broken pre-this-PR as well so I think we're good to go there.

File: crates/bevy_render/src/view/window.rs [hunk 9]
```diff
@@ -165,16 +165,15 @@ pub fn prepare_windows(
         .windows
         .values_mut()
         // value of raw_winndow_handle only None if synthetic test
```
Comment (rparrett) [inline crates/bevy_render/src/view/window.rs:167]:
```suggestion
        // value of raw_handle is only None in synthetic tests
```

**Comment by rparrett** [comment]
I got the android example running on my device (pixel 2 xl) but only with `NDK 25.1.8937393` and compiling without `bevy_audio`.
```
cargo apk run --example android_example --no-default-features --features "bevy_winit,render,bevy_asset"
```
I had to lower a `WgpuLimit`:
```rust
let mut settings = WgpuSettings {
    priority: WgpuSettingsPriority::Compatibility,
    ..default()
};
```
And resuming after putting the app in the background causes it to crash with this error:
```
10-19 12:52:26.837  8148  8181 E vulkan  : NATIVE_WINDOW_MIN_UNDEQUEUED_BUFFERS query failed: No such device (-19) value=0
10-19 12:52:26.837  8148  8181 E log event:  get_physical_device_surface_present_modes: ERROR_SURFACE_LOST_KHRlog.target = "wgpu_hal::vulkan::adapter"; log.module_path = "wgpu_hal::vulkan::adapter"; log.file = "/Users/robparrett/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-hal-0.14.0/src/vulkan/adapter.rs"; log.line = 1512;
10-19 12:52:26.838  8148  8181 E vulkan  : dequeueBuffer failed: No such device (-19)
```

**Review by rparrett — APPROVED** [review]

**Comment by VitalyAnkh** [comment]
> And resuming after putting the app in the background causes it to crash with this error


I think this deserves an issue report, and let's fix it in the future :rocket:

**Comment by mockersf** [comment]
> I had thought that we'd be able to do audio now that we have ndk-glue 0.7 across the board. But this was my first time actually succeeding in running on android and I haven't been following the progress there closely.

No, rodio hasn't been updated to the latest latest ndk-glue...  😞 

> Regardless, I don't think there's a regression here, so we're probably good to go.

Yup, all things you mentioned are expected 👍

**Comment by mockersf** [comment]
thanks @rparrett for testing on iOS and Android 🎉

**Pull request closed without merging**